### PR TITLE
[RyuJIT] Unroll single-iteration loops

### PR DESCRIPTION
**Pull request by EgorBo**
JIT is able to unroll small constant-size loops via `optUnrollLoops`, but currently it only works if `Vector_.Count` is used ([magic](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgDkBXfGKASzFwG4BYAKD+pJKAJnIBhPgG8+5WZSpDWAOwzkAMhAgAHKgH06EDAFUlUCABtzMACYAKAJQy503nLfllqhOQC85Ujyu7rIAZtDktp4evv6c0QA85AAccawA1GmOQcFuCBmBOZQA7OQIBXIAvk6y1eS1gh4q6po6uiZmljYOtS6FUd5+AbVuYVARUawxAQnkAGowYBjQ8Z4AfDRiEAxNAPQAVMnkezupGVmFufnDcsQlZbVVvBVAA=)) as the upper limit. This PR extends that behavior to also handle single iteration loops (thus it won't regress anything)

```csharp
static int GetCount() => 1; // e.g. static readonly field inited via an envvar

static int Test()
{
    int x = 0;
    for (int i = 0; i < GetCount(); i++)
        x++;
    return x;
}
```
### Current codegen for `Test`:
```asm
G_M58954_IG01:
G_M58954_IG02:
       xor      eax, eax
       xor      edx, edx
G_M58954_IG03:
       inc      eax
       inc      edx
       test     edx, edx
       jle      SHORT G_M58954_IG03
G_M58954_IG04:
       ret
```

### New codegen for `Test`:
```asm
G_M58954_IG02:
       mov      eax, 1
G_M58954_IG03:
       ret  
```

### jit-diff (`-f --cctors --pmi`):
(surprisingly, it's not empty)
```
Total bytes of delta: -352 (-0.00% of base)
    diff is an improvement.

Top file improvements (bytes):
        -211 : System.Private.CoreLib.dasm (-0.00% of base)
        -103 : Microsoft.CodeAnalysis.dasm (-0.01% of base)
         -38 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm (-0.00% of base)

3 total files with Code Size differences (3 improved, 0 regressed), 265 unchanged.

Top method improvements (bytes):
        -103 (-12.50% of base) : Microsoft.CodeAnalysis.dasm - SmallDictionary`2:LeftComplex(AvlNode):AvlNode (8 base, 7 diff methods)
         -66 (-12.69% of base) : System.Private.CoreLib.dasm - Vector64`1:Equals(Vector64`1):bool:this (6 methods)
         -66 (-11.21% of base) : System.Private.CoreLib.dasm - Vector64`1:GetHashCode():int:this (6 methods)
         -62 (-4.52% of base) : System.Private.CoreLib.dasm - Utf8Formatter:TryFormat(TimeSpan,Span`1,byref,StandardFormat):bool
         -17 (-5.28% of base) : System.Private.CoreLib.dasm - Utf8Parser:TryParseByteX(ReadOnlySpan`1,byref,byref):bool
         -14 (-24.56% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - FileVersionTraceData:get_FileDescription():String:this
         -13 (-1.53% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - FileVersionTraceData:ToXml(StringBuilder):StringBuilder:this
         -11 (-1.75% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - FileVersionTraceData:PayloadValue(int):Object:this

Top method improvements (percentages):
         -14 (-24.56% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - FileVersionTraceData:get_FileDescription():String:this
         -66 (-12.69% of base) : System.Private.CoreLib.dasm - Vector64`1:Equals(Vector64`1):bool:this (6 methods)
        -103 (-12.50% of base) : Microsoft.CodeAnalysis.dasm - SmallDictionary`2:LeftComplex(AvlNode):AvlNode (8 base, 7 diff methods)
         -66 (-11.21% of base) : System.Private.CoreLib.dasm - Vector64`1:GetHashCode():int:this (6 methods)
         -17 (-5.28% of base) : System.Private.CoreLib.dasm - Utf8Parser:TryParseByteX(ReadOnlySpan`1,byref,byref):bool
         -62 (-4.52% of base) : System.Private.CoreLib.dasm - Utf8Formatter:TryFormat(TimeSpan,Span`1,byref,StandardFormat):bool
         -11 (-1.75% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - FileVersionTraceData:PayloadValue(int):Object:this
         -13 (-1.53% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - FileVersionTraceData:ToXml(StringBuilder):StringBuilder:this

8 total methods with Code Size differences (8 improved, 0 regressed), 259114 unchanged.
Completed analysis in 23.84
```

Example: [Utf8Parser:TryParseByteX](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Unsigned.X.cs#L52) (`ByteOverflowLengthHex = 2`) - looks suspicious btw:

### CODE REVIEW

**Comment by EgorBo** [comment]
@dotnet/jit-contrib PTAL

**Comment by AndyAyersMS** [comment]
cc @BruceForstall

File: src/coreclr/src/jit/optimizer.cpp [hunk 1]
```diff
@@ -3647,6 +3635,29 @@ void Compiler::optUnrollLoops()
...
+        {
+            if (totalIter <= 1)
+            {
+                unrollLimitSz *= 4;
+            }
+            else
+            {
+                continue;
+            }
+        }
```
Comment (BruceForstall) [inline src/coreclr/src/jit/optimizer.cpp:3659]: Why do we increase the unroll limit in this case? Shouldn't we only do that in the stress mode?

Shouldn't this just be:
```
    // Unroll only if limit is 0, 1, or Vector_.Length (as a heuristic, not for correctness/structural reasons)
    if (!(loopFlags & LPFLG_SIMD_LIMIT) && (totalIter > 1))
    {
        continue;
    }
```

**Review by BruceForstall — COMMENTED** [review]
Question

File: src/coreclr/src/jit/optimizer.cpp (hunk 1 above)
Comment (EgorBo) [reply src/coreclr/src/jit/optimizer.cpp]: Could you please re-review the recent changes?

**Review by sandreenko — APPROVED** [review]
LGTM, thank you.

**Pull request merged**